Entities/AHuman.cpp

```diff
@@ -1796,6 +1803,39 @@ bool AHuman::UpdateMovePath()
     return true;
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+
+void AHuman::UpdateWalkAngle(AHuman::Layer whichLayer) {
+	if (m_Controller.IsState(BODY_JUMP)) {
+		m_WalkAngle[whichLayer] = Matrix(c_QuarterPI * GetFlipFactor());
+	} else {
+		// Cast rays to calculate the approximate shape of terrain.
+		int rayCount = 4;
+		float rayLength = 10.0F;
+		Vector hipPos = m_Pos;
+		if (whichLayer == AHuman::Layer::FGROUND && m_pFGLeg) {
+			rayLength += m_pFGLeg->GetMaxLength();
+			hipPos += RotateOffset(m_pFGLeg->GetParentOffset());
+		} else if (m_pBGLeg) {
+			rayLength += m_pBGLeg->GetMaxLength();
+			hipPos += RotateOffset(m_pBGLeg->GetParentOffset());
+		}
+		float traceRotation = rayCount > 1 ? c_HalfPI / static_cast<float>(rayCount - 1) * GetFlipFactor() : 0;
+		Vector hitPos;
+		Vector terrainVector(0, rayLength);
+		Vector trace(0, rayLength);
+		for (int i = 0; i < rayCount; i++) {
+			if (g_SceneMan.CastStrengthRay(hipPos, trace, 10.0F, hitPos, 4, g_MaterialGrass)) {
+				terrainVector += trace - g_SceneMan.ShortestDistance(hipPos, hitPos, g_SceneMan.SceneWrapsX());
+			} else {
+				// Reinforce focus on previously assumed terrain.
```

Comment:
If you wanna keep this comment, I'd like to be expanded a bit to be very clear in what it's saying. Maybe something like
```
Since no terrain was found, reinforce the existing terrain direction Vector, so any terrain found in future affects it less.
```